src/neo/SmartContract/Native/Designate/DesignateContract.cs

```diff
@@ -37,29 +38,64 @@ public ECPoint[] GetDesignatedByRole(StoreView snapshot, Role role)
         {
             if (!Enum.IsDefined(typeof(Role), role))
                 throw new ArgumentOutOfRangeException(nameof(role));
-            return snapshot.Storages[CreateStorageKey((byte)role)].GetInteroperable<NodeList>().ToArray();
+            byte[] index = snapshot.Storages[CreateStorageKey((byte)role)].Value;
+            KeyBuilder index_key = CreateStorageKey((byte)role).Add(index);
+            return snapshot.Storages[index_key].GetInteroperable<NodeList>().ToArray();
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        public ECPoint[] GetDesignatedByRoleAndIndex(StoreView snapshot, Role role, uint index)
+        {
+            if (!Enum.IsDefined(typeof(Role), role))
+                throw new ArgumentOutOfRangeException(nameof(role));
+            if (snapshot.Height + 2 < index)
+                throw new ArgumentOutOfRangeException(nameof(index));
+            List<uint> keys = snapshot.Storages.Find(CreateStorageKey((byte)role).ToArray())
+                .Select(p => p.Key.Key.Skip(1).ToArray())
+                .Where(p => p.Length == sizeof(uint))
```

Comment:
We can do the same without the `Skip` and It will be faster

Reply:
Done.